pkg/buildqueue/buildqueue.go

```diff
@@ -0,0 +1,360 @@
+	currentBuild *build
+
+	// nextBuild is the building information of the next build for this
+	// UUID
+	nextBuild *build
+}
+
+func (q *BuildQueue) enqueueBuild(b Builder) (BuildNotification, bool) {
+	build, enqueued := q.serializeBuild(b)
+
+	notify := newBuildNotification()
+	build.addNotificationChannel(notify)
+
+	return notify, enqueued
+}
+
+func (q *BuildQueue) serializeBuild(b Builder) (*build, bool) {
+	build := newBuildInfo(b)
+	uuid := b.GetUUID()
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if buildStatus, ok := q.buildStatus[uuid]; ok {
+		// If the builder is currently being built, prepare the next
+		// build and store it in the nextBuild field. When the current
+		// build finishes, the next build will automatically be added
+		// to the workerBuildQueue
+		if buildStatus.currentBuild != nil && buildStatus.nextBuild == nil {
+			buildStatus.nextBuild = build
+			return build, false
+		}
+
+		// The builder is already in the queue but not being built, the
+		// build request will be fulfilled when the queued build is
+		// executed. The caller can be notified to skip the build.
+		return buildStatus.nextBuild, false
+	}
+
+	q.buildStatus[uuid] = &buildStatus{nextBuild: build}
+
+	return build, true
+}
+
+// Enqueue schedules Builder for building. A channel is returned to provide the
+// ability to wait for the build to complete and check for success or failure.
+func (q *BuildQueue) Enqueue(b Builder) BuildNotification {
+	notify, enqueue := q.enqueueBuild(b)
+	if enqueue {
+		// This should be non-blocking unless there is contention beyond
+		// queueBuildSize in which case this will block until a slot in the
+		// queue becomes available.
+		q.workerBuildQueue <- b
+	}
+
+	return notify
+}
+
+// Remove removes the builder from the queue.
+func (q *BuildQueue) Remove(b Buildable) {
+	q.mutex.Lock()
+	delete(q.buildStatus, b.GetUUID())
+	q.mutex.Unlock()
+}
+
+// Drain will drain the queue from any ongoing or scheduled builds of a
+// Buildable. If a build is ongoing, the function will block for the build to
+// complete. It is the responsibility of the caller that the Buildable does not
+// get re-scheduled during the draining. Returns true if waiting was required.
+func (q *BuildQueue) Drain(b Buildable) bool {
+	uuid := b.GetUUID()
+
+	q.mutex.Lock()
+	status, ok := q.buildStatus[uuid]
+	var currentBuild *build
+	if ok {
+		currentBuild = status.currentBuild
+		delete(q.buildStatus, uuid)
+
+	}
+
+	q.mutex.Unlock()
+
+	// If a build is onging, block until build is complete
+	if currentBuild != nil {
+		<-currentBuild.internalNotification
+		return true
+	}
+
+	return false
+}
+
+func (q *BuildQueue) runExclusiveBuild(b Builder) {
+	uuid := b.GetUUID()
+
+	for {
+		currentBuild := q.dequeueBuild(uuid)
+		if currentBuild == nil {
+			return
+		}
+
+		// Register need for an exclusive build
+		q.ongoingExclusiveBuilds.Add(1)
+
+		// Wait for all regular builds to complete
+		q.ongoingRegularBuilds.Wait()
+
+		err := b.Build()
```

Comment:
I believe with the current code, this build is not guarantee to be exclusive. Having a sleep in [0] will trigger what I'm saying.

Why don't you replace both `WaitGroup` with a `RWMutex` and use the `Lock` for the `ongoingExclusiveBuilds` and the `RLock` for the `ongoingRegularBuilds`? Oh because you want the others to finish before triggering this one... Does that really matter?

Reply:
The whole point is to avoid holding mutexes for such a long time. This is still marked as wip. I'm not sure why everybody is reviewing this yet ;-) There are several other issues outstanding as well.

Reply:
OK, OK. Just hold your horses for a little bit longer ;-)